test/distributed/test_fake_pg.py

```diff
@@ -168,6 +175,73 @@ def test_recv(self):
         dist.recv(output, 1)
         self.assertEqual(tuple(output.shape), (3, 3))
 
+    @unittest.skipIf(not HAS_CUDA or not enable_2d_with_fsdp(), "No CUDA or TP+FSDP")
+    def test_fsdp_tp_sp_fake_e2e(self):
+        world_size = 4
+        tp_size = 2
+        parallel_style = SequenceParallel()
+
+        store = dist.HashStore()
+        dist.init_process_group(backend="fake", rank=0, world_size=world_size, store=store)
+
+        device_mesh = DeviceMesh(
+            "cuda", torch.arange(0, world_size).view(-1, tp_size)
+        )
+
+        my_module = nn.Sequential(
+            nn.Linear(10, 32),
+            nn.ReLU(),
+            nn.Linear(32, 5),
+        ).cuda(dist.get_rank())
+
+        my_module = parallelize_module(my_module, device_mesh, parallel_style, tp_mesh_dim=1)
+
+        sharded_module = FSDP(my_module, use_orig_params=True)
+        optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+
+        for i in range(10):
+            dp_rank = dist.get_rank()
+            torch.manual_seed(i + dp_rank)
+            input = torch.randn(20, 10).cuda(dist.get_rank())
+            x = sharded_module(input)
+            loss = x.sum()
+            loss.backward()
+            optim.step()
+
+    @unittest.skipIf(not HAS_CUDA or not enable_2d_with_fsdp(), "No CUDA or TP+FSDP")
+    def test_fsdp_tp_pp_fake_e2e(self):
```

Comment:
naming seems indicating we have "pp" but the test does not have pipeline parallel? Agree with @fduwjj that we should consolidate the SP/TP test case a bit, the only difference here looks to be the `parallel_style`, we can simply loop through parallel styles and verify each case works in a single test case

Reply:
pp was intended to refer to using `PairwiseParallel` instead of `SequenceParallel`, I'll change it since it's confusing

The other minor difference with SP test case is how we retrieve the `dp_rank`, but it can be combined into a loop as well